thoth/python/source.py

```diff
@@ -193,6 +182,78 @@ def _warehouse_get_package_hashes(
 
         return result
 
+    def _elf_find_versioned_symbols(self, elf: ELFFile) -> Iterator[Tuple[str, str]]:
+        """Takes an ELFFile object and outputs the required dynamic symbols."""
+        section = elf.get_section_by_name(".gnu.version_r")
+
+        if section is not None:
+            for verneed, verneed_iter in section.iter_versions():
+                if verneed.name.starstwith("ld-linux"):
+                    continue
+                for vernaux in verneed_iter:
+                    yield verneed.name, vernaux.name
+
+    def _is_elf(self, filename: str):
+        """Checks if files magic numbers are <delete>ELF."""
+        with open(filename, "rb") as f:
+            return f.read(16).startswith(b'\x7f\x45\x4c\x46')
+
+    def _get_versioned_symbols_from_file(self, filename: str):
+        """Given a file get all required dynamic symbols if it's an executable."""
+        to_ret = set()
+        with open(filename, "rb") as f:
+            if not _is_elf(filename):
+                return to_ret
+            elf = ELFFile(f)
+            for _, sym in self._elf_find_versioned_symbols(elf):
+                to_ret.add(sym)
+        return to_ret
+
+    def _get_versioned_symbols_in_dir(self, root: str):
+        """Walk dir and get all dynamic symbols required from all files"""
+        to_ret = set()
+        for dir_name, subdir_list, file_list in os.walk(root):
+            for fname in file_list:
+                to_ret = to_ret | self._get_versioned_symbols_from_file(os.path.join(dir_name, fname))
+
+        return to_ret
+
+    def _get_data_from_compressed(self, compressed_file: str, artifact_name: str):
+        """Run extract and call any functions on dir to get desired info"""
+        try:
+            tempdir = tempfile.mkdtemp()
+            try:
+                with zipfile.ZipFile(compressed_file, "r") as zip_ref:
+                    zip_ref.extractall(tempdir)
+                    to_ret["syms"] = self._get_versioned_symbols_in_dir(tempdir)
+                    to_ret["hashes"] = self._gather_hashes(tempdir)
+                    # NOTE: If you want to add more functions do it here
+                _LOGGER.debug("File is a .whl file")
+            except Exception as e:
+                tf = tarfile.open(compressed_file)
+                tf.extractall(tempdir)
+                to_ret["syms"] = self._get_versioned_symbols_in_dir(tempdir)
+                to_ret["hashes"] = self._gather_hashes(tmpdir)
```

Comment:
@KPostOffice maybe tempdir?

Reply:
From desing point of view, it could be better to introduce a new class `PythonArtifact` which would be returned by `Source.download_artifact` (or artifacts) and provide these method on the newly introduced class (e.g. `PythonArtifact.get_versioned_symbols_in_dir`, `PythonArtifact.gather_hashes`, ...) The solution would be extensible when it comes to new analyses on top of PythonArtifacts - later on we can gather additional attributes such as wheel tags and so on.

Reply:
Thanks, copy paste mistake :no_mouth:

Reply:
The introduced PythonArtifact would carry information about the downloaded artifact and it would clean all the extracted files (tempdir) on destruction.

Reply:
I like this idea

Reply:
Just... As we are in `thoht-python` which groups Python related stuff, `Artifact` class name would be more appropriate :)